### ISSUE TITLE
Make Code Style Guide for Numberscope

### ISSUE THREAD

**Issue by liammulh** [numberscope/frontscope#69]
@katestange, @gwhitney, before I add the Prettier code formatter, it would be useful to get your opinions on what our TypeScript code style / code format should be going forward.

First, I highly recommend skimming Idiomatic JavaScript (https://github.com/rwaldron/idiomatic.js/), a fairly popular guide for JavaScript code style. We obviously don't need to agree with all of their rules, but I think they have a good approach/philosophy when it comes to code style and code format.

Second, would you please comment as to whether your text editor or IDE has a plugin for Prettier? See their supported text editors and IDE's here: https://prettier.io/docs/en/editors.html.

If one of you uses a text editor that isn't supported, we could try using a combination of https://editorconfig.org/ and ESLint rules. Otherwise, to familiarize yourselves with some of the options Prettier provides, please click this link: https://prettier.io/playground/.

Some of the big things we need to decide on:

1. still going with 80-character limit?
2. tabs or spaces? width?
3. single quote or double quote?
4. trailing comma?
```js
const myTrailingCommaObj = {
  a: 1,
  b: 2,
  c: 3,
}

const myRegularObj = {
  a: 1,
  b: 2,
  c: 3
}
```
5. semicolons?
6. conditional style?
```js
if(condition) doSomething();

if(condition) {
  doSomething();
}
```
7. do we want to abbreviate variable names or be verbose?

I'm sure there are other things as well. If you can think of any, please write them in this issue. If we can't use Prettier to enforce some things, I bet we can find an ESLint rule.

Alternatively, we could just use ESLint rules for formatting, but the downside to this is that you wouldn't be able to use your editor to "autoformat" your code, which I find saves a lot of time.

Finally, I think we should create a living code style guide that we can add to / subtract from as needed.

**Comment by liam-mulhall** [comment]
We also need to agree on a style for conditionals. For example, [see this PR conversation](https://github.com/numberscope/frontscope/pull/51/files/ab053262e5030fab31fcfb9613f6bf3476d0bba6#diff-52110c64f50a9009e3b18d7c6507ba770c193c70c7ee04cfe3b7ad1b08ee8d6c) between me and Glen.

**Comment by gwhitney** [comment]
I use emacs, which has a prettier mode, so we are cool there. I will read your recommended references and weigh in on the issues you raise. Where should the code style guide live? In the docs within the repository, or here in the repository manager (github), say as part of the PR guidelines?

**Comment by liam-mulhall** [comment]
> Where should the code style guide live?

I think it makes sense to keep a guide as a markdown file in a `doc` directory in the Git repository rather than using GitHub's wiki feature or something similar. While I like GitHub's wiki feature, I think that docs should be in version control and should be available locally if you've cloned the repo. If there's a feature where we can have a doc directory and have that auto-populate the wiki, that would be the best of both worlds.

**Comment by liam-mulhall** [comment]
I just read the docs about GitHub wikis, and it seems like the wiki is sort of separate from the Git repository? You'd have to clone both the repository and the wiki to get the wiki locally if I'm reading the docs correctly.

https://docs.github.com/en/communities/documenting-your-project-with-wikis/adding-or-editing-wiki-pages

**Comment by katestange** [comment]
I use neovim, which supports a plugin, and my window is usually at least 100 characters wide while I'm working.  As for code styles, I think that is best left for you and Glen to hash out.  I don't have strong or informed opinions.  The only thing I find myself having any opinion on is that I like to have readableVariableNames (not abbrvVNs) -- if that's what you're asking.  It helps me understand the code.

**Comment by gwhitney** [comment]
> I just read the docs about GitHub wikis, and it seems like the wiki is sort of separate from the Git repository? You'd have to clone both the repository and the wiki to get the wiki locally if I'm reading the docs correctly.
> 
> https://docs.github.com/en/communities/documenting-your-project-with-wikis/adding-or-editing-wiki-pages

Splitting this side conversation into a separate issue, so that this issue can focus on the content of the style guide, which appears to be its main thread.

**Comment by gwhitney** [comment]
My most-comfortable approaches to the seven particular aspects Liam brought up (per his numbering above):

1) I can get two 80-char windows side-by-side on my screen (but not really larger line lengths in a font I can read), so I like 80-character limit. That said, if there is a consensus for 96 or something, I am happy to go along with that.

2) I think we should use all-spaces for indentation. Then it looks the same to everyone, no matter what kind of viewer they use.

3) For literal string constants, I prefer single quotes since from my Bash-shell days, they "look" like no substitution or anything will occur. Then for TypeScript, we have backquotes for strings with expression interpolation, and we basically would not use double quotes at all.

4) No opinion - fine either way. Liam, if you are the only one with a view on this, feel free to go ahead and pick.

5) My understanding is that semicolons are completely unnecessary, or at least very very close to that, in TypeScript, so I'd suggest we not use them -- except of course in traditional for-loops, where they are part of the syntax.

6) I've already commented on this in #51 -- I feel both conditional styles should be tolerated, with the "one-liner" used judiciously to make code more vertically compact and easier to read, when it does that, otherwise the brace style should be used. What I feel strongly should be disallowed are both:
```
if (someCondition) { oneLiner(withBraces) == 'why did you put braces in?' }
```
and
```
if (bigLongCondtionOrSemanticallyImportantChoice)
    breakLinesWithoutBraces == 'tempting fate to add another line below that won't be just in true path'
```

7) Also in #51 -- I feel that we should use self-explanatory, full-word variable names except for things like loop indices and "temporary" variables with textually very small scope -- the latter are often introduced for the sake of abbreviating long expressions (perhaps in order to comply with the line-length limit), and so short names should thus be allowed and shouldn't cause confusion, as there won't be any question of what the variable "means" if it is only in scope in the line or two following its definition.

**Comment by gwhitney** [comment]
Also in #51, Liam raised another issue:

8) What should the proper spacing around punctuation (parentheses, braces, commas, colons) be?

The coding guide should specify an answer to (8). My personal comfort lies at "the same as in standard written English" except for a couple of specific syntaxes like `functionCall(arg1, arg2)` where standard English would have a space between functionCall and the left parenthesis, which would break up the function call in a weird way in code.

**Comment by gwhitney** [comment]
Finally, although I agree that Idiomatic JavaScript (IJS) is generally a nice guide, we are writing TypeScript so we may want some different conventions as a result. Also there are a couple of recommendations in IJS that I wanted to go on record as  _disagreeing_ with, namely:

2.B.1: IJS recommends only declaring variables at the top of their scope. I feel strongly that variables should be introduced where they are needed, and not sooner (except for example in cases where several related variables are introduced together for clarity, even though one isn't used for a while).

3.B.1: IJS recommends type coercion with such constructs as prefix `+`. I think TypeScript supersedes such things.

4.1.1: IJS recommends testing "truthiness" `if (array.length) {...` rather than an explicit comparison `if (array.length > 0) {...` To me, the intent of the latter is crystal clear, and I don't see any problem with it, so I think it should be allowed.

7.A.1: IJS recommends against the switch statement altogether. I don't understand why, and there doesn't seem to be a good explanation. I definitely feel switch is the clearest way to express the desired behavior in some circumstances, and is generally better than a long sequence of conditionals when it can be substituted for one.

9: IJS recommends against end-of-line comments. If you have just a few words of explanation about what happens _on that line_, I like the greater compactness of end-of-line comments over /*... */ on the prior line, and don't see the problem with putting in a brief //-comment. I do try to avoid continuing an end-of-line comment to another line with another //, although I have found myself doing it -- I will be more consistent about eschewing that.

**Comment by gwhitney** [comment]
Adding item
10) Should all source files end in a newline?

I vote yes. Many tools expect all files to be newline-terminated, and I don't really know of any that are broken by the last line of a file ending in newline.

**Comment by gwhitney** [comment]
Once #90 is merged, the code style guide needs to mention Prettier and give guidance on how it is used, affects development, etc.

**Comment by gwhitney** [comment]
Raising additional style guide point:
11) Number of spaces per indent level?

My personal vote would be either for the current 4 or if people prefer less, 3; I find 2 too few to easily follow visually.

**Comment by gwhitney** [comment]
And most recently from the discussion in #90:
12) Line breaks before or after operators when necessary?
In other words,
```
let answer = veryLongSubexpression
    + anotherRemarkablyLoquaciousTerm
```
or
```
let answer = veryLongSubexpression +
    anotherRemarkablyLoquaciousTerm
```

My personal vote is the first option, before, because it's much easier to scan the left edge of the code to see the overall
structure of the expression than to find the operators at the right ends of the lines, which are much more likely to be highly varying in horizontal location.

**Comment by gwhitney** [comment]
A point just came up about code formatting issue no 1., the 80-character line limit. We need to decide whether that applies to `//` comments as well, especially as `prettier` it seems does not enforce this either way. So it will just be up to our own conventions. (Personally,  I like a uniform line length limit so I never need to decipher wrapped lines, but I will go with whatever the consensus is.)

**Comment by gwhitney** [comment]
On point 1 (line length) the point arose in discussion today whether the limit should apply to comment lines. The consensus appeared to be that our convention should be that it "almost always" should; occasionally there might be URLs or other unbreakable text that we would allow to hang over the 80-character limit.

Note that prettier-eslint (see #90) is not able to enforce this or reformat comments. So the code conventions that are automatically satisfied by running prettier-eslint all the time versus those that need manual adherence need to be noted in the eventual style guide.

Nevertheless it will be good to check if we can figure the eslint part to at least warn if any lines are over 80 characters, even comment lines. If we can, I would recommend turning that on in our merge-review checks, so that at least there will be warnings when it occurs (and I personally am fine if any necessary violations have to be accompanied by a tag telling eslint that this one is OK.

**Comment by gwhitney** [comment]
Now that prettier-eslint is implemented and the associated format-check is required, that will take care of many of these items automatically. But note they are not a panacea; it's still necessary to look at new/changed code with an eye to format/style conventions. In particular, at least with vue files, our automatic tools will accept a variety of formatting/line breaks for the same code, so we need to choose something that looks the most reasonable. For example, in a `<div>...</div>` block in a template, prettier-eslint will accept the closing tag either on its own line aligned with the opening tag, or at the end of the last line of the
block content. So I have generally been putting it on its own line when there is multi-line content or it doesn't fit at the end
of the line because of the length cap, and otherwise putting it on the same line as the content. (Should something like that be in our style guide?)

**Comment by liam-mulhall** [comment]
Related to https://github.com/numberscope/frontscope/issues/127.

**Comment by liam-mulhall** [comment]
I don't think @gwhitney or @katestange need to supply more input, so I'm un-assigning them.

**Comment by liammulh** [comment]
Let's add an item for return statements.

- If you can easily figure out that you need to bail out of a function, go ahead and put a return at the top of the function and then proceed with the rest of the function.
- Otherwise, try to minimize the number of return statements in a function. Having them strewn throughout a function (especially a long function) is generally hard to read.

**Comment by liammulh** [comment]
I don't have anything else to add at the moment. In the future, I think we should just open a draft PR when we want to add something to the code style doc. Closing.
